wrappers/core18.go

```diff
@@ -657,6 +661,35 @@ func undoSnapdDbusActivationOnCore() error {
 	return err
 }
 
+var snapdDesktopFileNames = []string{
+	"io.snapcraft.SessionAgent.desktop",
+	"snap-handle-link.desktop",
+}
+
+func writeSnapdDesktopFilesOnCore(s *snap.Info) error {
+	// Ensure /var/lib/snapd/desktop/applications exists
+	if err := os.MkdirAll(dirs.SnapDesktopFilesDir, 0755); err != nil {
+		return err
+	}
+
+	desktopFiles := make(map[string]osutil.FileState, len(snapdDesktopFileNames))
+	for _, fileName := range snapdDesktopFileNames {
+		filePathInSnap := filepath.Join(s.MountDir(), "usr/share/applications", fileName)
+		if !osutil.FileExists(filePathInSnap) {
+			continue
```

Comment:
given that the code/list and the files should come from the same place, shouldn't this be an error or at least be logged?

Reply:
The code is like this in response to this discussion: https://github.com/snapcore/snapd/pull/13078#discussion_r1319500816

Consider the case where snapd removes a desktop file that it previously shipped. We'd want to keep the file name in `snapdDesktopFileNames` even though it doesn't exist in the new snapd snap, so that the EnsureDirStateGlobs() call removes the obsolete file on upgrade.